ts/plugins/importFlake.ts

```diff
@@ -0,0 +1,46 @@
+import { Check } from "../check.ts";
+import { Executable, mkExecutable } from "../executable.ts";
+import { mapValues } from "../internal/utils.ts";
+import { nixFlakeDep, nixRaw, nixStrLit } from "../nix.ts";
+import { Plugin } from "../project.ts";
+
+export function importFlake<
+  Apps extends Record<string, string>,
+  Checks extends Record<string, string>,
+>(
+  flakeDir: string,
+  toImport: {
+    apps: Apps;
+    checks: Checks;
```

Comment:
Interesting. I was thinking that this would somehow create records for `Package`s, `Check`s & `Executable`s with dynamic (i.e. not statically known) keys. But this is of course kind of nice. Although, maybe sometimes you just want to import a flake and ignore type-safety? Maybe we want to allow both? So maybe this should add all the given fields to the project in a type safe way, but *also* add everything else in an unsafe way somehow?

In any case, why does `Apps` extend `Record<string, string>`? Is the key the name under which it'll be added to the project and the value is the name under which it's known in the flake file?

Reply:
I also hadn't thought of it, but I like the explicit and typed solution.

Not sure I like it being a `Plugin`, though. The fact that the `project` argument is ignored is, in my mind, evidence that it shouldn't be a plugin. "Lifting" it to a `Plugin` is always trivial. And there are other patterns of usage (such as just picking a few entries to add to one project, and a few to another) that also make sense.

If it's *not* a plugin, I think it also doesn't make sense to do name remapping in the import itself. I'm imagining usage being something like:

```typescript
const flake = importFlake({
  apps: ["app1", "app2"],
  checks: ["formatting", "tests"],
  packages: ["backend"] 
})

export const foo = someproj.add((_) => { 
  format: flake.checks.formatting, 
  server: flakes.packages.backend
 })

export const foo = someproj.add((_) => { 
  ci: flake.checks.tests
  devServer: flake.checks.app1
 })
```


But the more important message: this is overall WAY COOL!

Reply:
> But the more important message: this is overall WAY COOL!

Agreed!

I think one use-case that we should definitely support is to get *all* `Check`s, `Package`s, `Executable`s & `Environment`s from the imported flake file and allow to *somehow* expose them to `garn`. If someone ends up adding e.g. a `Check` to an imported flake file it'd be really sad if that `Check` got silently ignored on e.g. CI.

Reply:
Time for a `noUnusedFlakeExports` check/plugin :)

Reply:
Here's an idea for a possible api:

```ts
// allChecks is a check that passes if all checks from the subflake pass.
// allPackages is a derivation where all packages from the subflake are in subdirectories.
export const subFlake: { allChecks: garn.Check, allPackages: garn.Package} = importFromFlake(".garn/flake.nix");

// Here we promise that `whatever` exists.
// This will cause a runtime error not when executing the garn file, but only when trying to run the `Executable`.
const whatever: garn.Executable = subFlake.getExecutable("whatever");

const someCheck: garn.Check = subFlake.getCheck("someCheck");
```

This API would work without calling e.g. `nix flake show --json` when executing the garn file.

Reply:
Although calling `nix flake show --json` may be the right call here. It'll give much better help messages. And tab-completion may get slower, but also would be more accurate.